### PR TITLE
feat: enable tagging manifests

### LINKED ISSUE #471: Command to create a tag for a manifest in the remote registry
The proposed command is:

```
Usage:
  oras tag  REPO:<TAG|DIGEST> NEW_TAG

Flags:
  --ca-file string               server certificate authority file for the remote registry
  -d, --debug                    debug mode
  -h, --help                     help for tagging
      --insecure                 allow connections to SSL registry without certs
  -p, --password string          registry password or identity token
      --password-stdin           read password or identity token from stdin
      --plain-http               allow insecure connections to registry without SSL check
  -u, --username string          registry username
  -v, --verbose                  verbose output
```

Relate to #459

### PR DESCRIPTION
**Pull request by wju-MSFT**
Resolve #471 

Signed-off-by: Jun <wju@microsoft.com>

### CODE REVIEW

File: cmd/oras/tag/tag.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	"oras.land/oras/cmd/oras/internal/errors"
+	"oras.land/oras/cmd/oras/internal/option"
+)
+
+type tagOptions struct {
+	option.Common
+	option.Remote
+
+	targetRef string
+	tagRef    string
```
Comment (Wwwsylvia) [inline cmd/oras/tag/tag.go:31]: nit: The naming looks a bit confusing to me as I can't tell which is the original ref and which is the new one. Would something like `srcRef`, `dstRef` be clearer?

File: cmd/oras/tag/tag.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+		return errors.NewErrInvalidReference(repo.Reference)
+	}
+
+	ref := opts.targetRef
+	desc, err := repo.Resolve(ctx, ref)
+	if err != nil {
+		return err
+	}
+
+	repo.Tag(ctx, desc, opts.tagRef)
```
Comment (Wwwsylvia) [inline cmd/oras/tag/tag.go:77]: You may want to leverage `oras.Tag`, which is convenient and performance optimized.

Comment (qweeah) [reply cmd/oras/tag/tag.go:77]: I think you are referring to [repository.TagReference](https://pkg.go.dev/oras.land/oras-go/v2@v2.0.0-rc.2/registry/remote#Repository.TagReference)?

File: cmd/oras/tag/tag.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+func TagCmd() *cobra.Command {
+	var opts tagOptions
+	cmd := &cobra.Command{
+		Use:   "tag name[:tag|@digest] new_tag",
+		Short: "[Preview] tag a manifest in the remote registry",
+		Long: `[Preview] tag a manifest in the remote registry
+** This command is in preview and under development. **
+
+Example -  Tag a manifest with tag v1.0.1 to repository 'locahost:5000/hello' and tag with 'v1.0.2':
+oras manifest push localhost:5000/hello:v1.0.1 v1.0.2
```
Comment (qweeah) [inline cmd/oras/tag/tag.go:43]:
```suggestion
oras manifest tag localhost:5000/hello:v1.0.1 v1.0.2
```

File: cmd/oras/tag/tag.go [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+func TagCmd() *cobra.Command {
+	var opts tagOptions
+	cmd := &cobra.Command{
+		Use:   "tag name[:tag|@digest] new_tag",
+		Short: "[Preview] tag a manifest in the remote registry",
+		Long: `[Preview] tag a manifest in the remote registry
+** This command is in preview and under development. **
+
+Example -  Tag a manifest with tag v1.0.1 to repository 'locahost:5000/hello' and tag with 'v1.0.2':
```
Comment (qweeah) [inline cmd/oras/tag/tag.go:42]:
```suggestion
Example -  Tag a manifest 'v1.0.1' in repository 'locahost:5000/hello' to 'v1.0.2':
```

File: cmd/oras/tag/tag.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+	option.Remote
+
+	targetRef string
+	tagRef    string
+}
+
+func TagCmd() *cobra.Command {
+	var opts tagOptions
+	cmd := &cobra.Command{
+		Use:   "tag name[:tag|@digest] new_tag",
```
Comment (qweeah) [inline cmd/oras/tag/tag.go:37]:
```suggestion
		Use:   "tag name<:tag|@digest> new_tag",
```

File: cmd/oras/tag/tag.go (hunk 2 above)
Comment (Wwwsylvia) [reply cmd/oras/tag/tag.go:77]: No I meant `oras.Tag`.

File: cmd/oras/tag/tag.go (hunk 4 above)
Comment (sajayantony) [reply cmd/oras/tag/tag.go:42]: Retap might be incorrect since you can also pass a ref with digest.

File: cmd/oras/tag/tag.go [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+		return err
+	}
+
+	if repo.Reference.Reference == "" {
+		return errors.NewErrInvalidReference(repo.Reference)
+	}
+
+	oras.Tag(ctx, repo, opts.srcRef, opts.targetRef)
+
+	return nil
```
Comment (qweeah) [inline cmd/oras/tag/tag.go:74]:
```suggestion
	return oras.Tag(ctx, repo, opts.srcRef, opts.targetRef)
```

**Review by qweeah — CHANGES_REQUESTED** [review]

**Comment by qweeah** [comment]
@junczhuMSFT Need to resolve the DCO

**Review by qweeah — APPROVED** [review]
LGTM

File: cmd/oras/tag/tag.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+func TagCmd() *cobra.Command {
+	var opts tagOptions
+	cmd := &cobra.Command{
+		Use:   "tag name<:tag|@digest> new_tag",
+		Short: "[Preview] tag a manifest in the remote registry",
+		Long: `[Preview] tag a manifest in the remote registry
+** This command is in preview and under development. **
+
+Example -  Tag the manifest 'v1.0.1' in 'locahost:5000/hello' to 'v1.0.2':
+oras tag localhost:5000/hello:v1.0.1 v1.0.2
```
Comment (shizhMSFT) [inline cmd/oras/tag/tag.go:42]: The indentation / spaces for examples seems inconsistent with other commands.

**Review by shizhMSFT — CHANGES_REQUESTED** [review]

File: cmd/oras/tag/tag.go [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+func TagCmd() *cobra.Command {
+	var opts tagOptions
+	cmd := &cobra.Command{
+		Use:   "tag name<:tag|@digest> new_tag",
+		Short: "[Preview] tag a manifest in the remote registry",
+		Long: `[Preview] tag a manifest in the remote registry
+** This command is in preview and under development. **
+
+Example -  Tag the manifest 'v1.0.1' in 'locahost:5000/hello' to 'v1.0.2':
```
Comment (shizhMSFT) [inline cmd/oras/tag/tag.go:41]: An example of tagging a manifest with digest will be perfect. e.g.
```
oras tag localhost:5000/hello@sha256:xxxxx v2
```

File: cmd/oras/tag/tag.go [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+	repo, err := opts.NewRepository(opts.srcRef, opts.Common)
+	if err != nil {
+		return err
+	}
+
+	if repo.Reference.Reference == "" {
+		return errors.NewErrInvalidReference(repo.Reference)
+	}
+
+	return oras.Tag(ctx, repo, opts.srcRef, opts.targetRef)
```
Comment (shizhMSFT) [inline cmd/oras/tag/tag.go:69]: We probably need to consider the output when tagging multiple tags. /cc @FeynmanZhou @yizha1 

This change will be in another PR.

Comment (wju-MSFT) [reply cmd/oras/tag/tag.go:69]: ack.

File: cmd/oras/tag/tag.go (hunk 7 above)
Comment (wju-MSFT) [reply cmd/oras/tag/tag.go:42]: removed extra space between "-" and "Tag".  formatted the indentation.

File: cmd/oras/tag/tag.go (hunk 9 above)
Comment (FeynmanZhou) [reply cmd/oras/tag/tag.go:69]: Yes, we need to provide output msg for different kinds of tagging results. @wju-MSFT Let's discuss and confirm it.

File: cmd/oras/tag/tag.go (hunk 8 above)
Comment (wju-MSFT) [reply cmd/oras/tag/tag.go:41]: Added an example for manifest with digest.

**Review by shizhMSFT — APPROVED** [review]
LGTM. @wju-MSFT The PR needs to be rebased to be merged.

**Comment by wju-MSFT** [comment]
> LGTM. @wju-MSFT The PR needs to be rebased to be merged.

@shizhMSFT PR is rebased and ready to be merged.

**Comment by wju-MSFT** [comment]
@shizhMSFT PR is rebased and ready to be merged.

**Review by shizhMSFT — APPROVED** [review]
LGTM
